pkg/controllermanager/controller/shoot/migration/reconciler.go

```diff
@@ -0,0 +1,116 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package migration
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	"github.com/gardener/gardener/pkg/controllermanager/apis/config"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/utils/kubernetes/health"
+)
+
+// Reconciler reconciles Shoots that should be prepared for the migration and sets a constraint when the destination
+// seed is ready.
+type Reconciler struct {
+	Client client.Client
+	Config config.ShootMigrationControllerConfiguration
+}
+
+// Reconcile reconciles Shoots that should be prepared for the migration and sets a constraint when the destination
+// seed is ready.
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := logf.FromContext(ctx)
+
+	ctx, cancel := controllerutils.GetMainReconciliationContext(ctx, controllerutils.DefaultReconciliationTimeout)
+	defer cancel()
+
+	shoot := &gardencorev1beta1.Shoot{}
+	if err := r.Client.Get(ctx, request.NamespacedName, shoot); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	patch := client.MergeFrom(shoot.DeepCopy())
+
+	if shoot.Status.LastOperation != nil && shoot.Status.LastOperation.Type == gardencorev1beta1.LastOperationTypeMigrate {
```

Comment:
We should check `ShouldPrepareShootForMigration` here as well. Otherwise there is a chance that the controller never removes the constraint if it misses the migration phase, i.e. when `shoot.Status.LastOperation.Type == gardencorev1beta1.LastOperationTypeMigrate`.

@plkokanov or @oliver-goetz could you double check please? I will push the change if you agree.

Reply:
Puh, I cannot say much about this tbh. 😅